File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.js
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
var connect = function(root) {
    if (!root) return null;
    const queue = [root];

    while (queue.length > 0){
        const rowLength = queue.length;

        for (let i = 0; i< rowLength; i++){
            const curr = queue.shift();
            if (i < rowLength - 1){
                curr.next = queue[0];
            }

            if (curr.left) queue.push(curr.left);
            if (curr.right) queue.push(curr.right);
        }
    }

    return root
};

File: 117-populating-next-right-pointers-in-each-node-ii/populating-next-right-pointers-in-each-node-ii.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root: return None
        queue = deque([root])

        while len(queue) > 0:
            rowLength = len(queue)

            for i in range(rowLength):
                curr = queue.popleft()
                if i < rowLength - 1:
                    curr.next = queue[0]

                if curr.left: queue.append(curr.left)
                if curr.right: queue.append(curr.right)

        return root